File: Programs/Prog13_01/LinkedList.java

public class LinkedList implements ListInterface {

    public class Node {

        private int info;
        private Node next;
        // private int length;

        public Node() {
            info = 0;
            next = null;
        }

        public void setInfo(int i) {
            info = i;
        }

        public void setNext(Node l) {
            next = l;
        }

        public int getInfo() {
            return info;
        }

        public Node getNext() {
            return next;
        }
    }

    private Node first;

    public LinkedList() {
        first = new Node();
    }

    public boolean isEmpty() {
        return (first.getNext() == null);
    }

    public void display() {
        Node current = first.getNext();

        while (current != null) {
            System.out.print(current.getInfo() + " ");
            current = current.getNext();
        }

        System.out.println();
    }

    public boolean search(int x) {
        Node current = first.getNext();

        while (current != null) {
            if (current.getInfo() == x) {
                return true;
            }
            current = current.getNext();
        }

        return false;
    }

    public void add(int x) {
        Node p = new Node();

        p.setInfo(x);
        p.setNext(first.getNext());

        first.setNext(p);
    }

    public void add(int[] nums) {
        for (int i = 0; i < nums.length; i++)
            add(nums[i]);
    }

    /**
     * Removes the first occurence of "x" in the list
     */
    public void remove(int x) {
        Node old = first.getNext();
        Node p = first;

        boolean found = false;
        while (old != null) {
            if (old.getInfo() == x) {
                found = true;
                break;
            } else {
                p = old;
                old = p.getNext();
            }
        }

        if (found) {
            p.setNext(old.getNext());
        }
    }

    /**
     * Insert value 'x' into a specific index 'loc'
     */
    public void insert(int x, int loc) {
        Node current = first.getNext();

        int i = 0;
        while (current != null) {
            i++;
            System.out.println("i: " + i);
            if (i - 1 == loc) {
                Node p = new Node();
                p.setInfo(x);
                p.setNext(current.getNext());
                current.setNext(p);
                return;
            }

            current = current.getNext();
        }

        System.out.println("Invalid location!");
    }

    // TODO: Implement
    public void removeItemAt(int loc) {

    }

    public int length() {
        Node current = first.getNext();
        int length = 0;

        while (current != null) {
            length++;
            current = current.getNext();
        }

        return length;
    }

    public void clear() {
        first = new Node();
    }

    @Override
    public String toString() {
        Node current = first.getNext();
        String output = "";

        while (current != null) {
            output += current.getInfo() + " ";

            current = current.getNext();
        }

        return output;
    }
}

File: Python/LinkedList.py
def main():
    linked_list = LinkedList()
    linked_list.add(1)
    linked_list.add(2)
    linked_list.add(3)
    linked_list.add(3)
    linked_list.add(4)
    linked_list.add(5)
    linked_list.add(6)
    linked_list.add(7)
    linked_list.add(8)

    linked_list.remove(3)
    linked_list.remove(4)
    linked_list.remove(8)
    linked_list.remove(7)
    linked_list.remove(1)

    linked_list.insert(99, 1)

    print(linked_list)

    linked_list.reverse()
    print(linked_list)

    linked_list.add(66)
    linked_list.insert(77, 3)
    print(linked_list)

    print(f"Found 99? {linked_list.search(99)}")
    print(f"Found 7? {linked_list.search(7)}")

    print(f'List size: {linked_list.size}')

    linked_list.clear()
    print(f"Is the list empty? {linked_list.is_empty()}")
    print(linked_list)


class Node:
    def __init__(self, data: int = None, next=None):
        self.data = data
        self.next = next


class LinkedList:
    def __init__(self):
        self.head = Node()
        self.size = 0

    def add(self, data: int) -> None:
        next = self.head.next
        self.head.next = Node(data, next)
        self.size += 1

    def insert(self, data: int, index: int) -> bool:
        if index > self.size:
            return False

        current = self.head.next
        prev = self.head

        i = 0
        while current:
            if i == index:
                next = prev.next
                prev.next = Node(data, next)
                self.size += 1
                return True

            i += 1
            prev = current
            current = current.next

        return False

    def remove(self, target: int) -> bool:
        if self.is_empty():
            return False

        current = self.head.next
        prev = self.head

        while current:
            if current.data == target:
                prev.next = current.next
                self.size -= 1
                return True

            prev = current
            current = current.next

        return False

    def search(self, target: int) -> bool:
        current = self.head.next

        while current:
            if current.data == target:
                return True

            current = current.next

        return False

    def reverse(self) -> None:
        current = self.head.next
        prev = None

        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next

        self.head = prev
        self.add(prev.data)
        self.head.data = None

    def is_empty(self) -> bool:
        return self.size == 0

    def clear(self) -> None:
        self.head = Node()
        self.size = 0

    def __str__(self) -> str:
        if self.is_empty():
            return '[]'

        output = '['

        current = self.head
        while current:
            if current.data:
                output += f'{current.data} -> '
            current = current.next

        return output + ']'


if __name__ == '__main__':
    main()
